### ISSUE TITLE
Panzer:  Negative Tpetra map indices for HDiv fields on meshes exceeding int32 limits

### ISSUE THREAD

**Issue by dridzal** [trilinos/Trilinos#10029]
## Bug Report
@trilinos/panzer

### Description
Tpetra throws an error due to negative indices for HDiv fields on periodic meshes exceeding int32 limits.  This is major blocker.  The error is 

`Tpetra::Map constructor (noncontiguous): Minimum global ID = -3630 over all process(es) is less than the given indexBase = 0.`

The error occurs on meshes containing indices > 2 billion (approximately), where periodic boundary conditions are enforced.  It is interesting that the HCurl and HGrad cases work.  Only HDiv fails.  This is likely another integer type error in the periodic BC matcher. 

### Steps to Reproduce
1. Compile the test code below.
2. Contact @dridzal to provide reproducer meshes.  These are small meshes, 10x10x10 elements.  We have two versions, one where entity indices start at zero, and another where they start at 7 billion.  The code will run successfully on the first mesh, and it will fail on the second mesh.
3. Observe the Tpetra error (partial output is below the code).

Code:

```
#include "Teuchos_GlobalMPISession.hpp"
#include "Teuchos_RCP.hpp"
#include "Teuchos_ParameterList.hpp"

#include "Kokkos_Core.hpp"

#include "Intrepid2_Basis.hpp"
#include "Intrepid2_HGRAD_HEX_C1_FEM.hpp"
#include "Intrepid2_HCURL_HEX_I1_FEM.hpp"
#include "Intrepid2_HDIV_HEX_I1_FEM.hpp"
#include "Intrepid2_HVOL_C0_FEM.hpp"

#include "Panzer_STK_MeshFactory.hpp"
#include "Panzer_STK_Interface.hpp"
#include "Panzer_STK_ExodusReaderFactory.hpp"
#include "Panzer_STKConnManager.hpp"
#include "Panzer_IntrepidFieldPattern.hpp"
#include "Panzer_DOFManager.hpp"

using namespace std;
using Teuchos::RCP;
using Teuchos::rcp;

int main(int argc, char * argv[]) {
  
  TEUCHOS_TEST_FOR_EXCEPTION(argc==1,std::runtime_error,"Error: this test requires a mesh file");
  
  Teuchos::GlobalMPISession mpiSession(&argc, &argv,0);
  Teuchos::RCP<Teuchos::MpiComm<int>> Comm = Teuchos::rcp( new Teuchos::MpiComm<int>(MPI_COMM_WORLD) );
  
  Kokkos::initialize();
  
  typedef Teuchos::RCP<const shards::CellTopology> topo_RCP;
  typedef Teuchos::RCP<Intrepid2::Basis<PHX::Device::execution_space, double, double > > basis_RCP;
  
  {
    
    // ==========================================================
    // Create a mesh from the file defined by the user
    // ==========================================================
    
    std::string input_file_name = argv[1];
    RCP<Teuchos::ParameterList> pl = rcp(new Teuchos::ParameterList);
    
    Teuchos::RCP<panzer_stk::STK_MeshFactory> mesh_factory = Teuchos::rcp(new panzer_stk::STK_ExodusReaderFactory());
    pl->set("File Name",input_file_name);
    
    // Optional: include periodic BCs
    if (argc>2) {
      RCP<Teuchos::ParameterList> pbc = rcp(new Teuchos::ParameterList);
      pbc->set("Count",2);
      pbc->set("Periodic Condition 1","xz-all 1e-8: top;bottom");
      pbc->set("Periodic Condition 2","yz-all 1e-8: left;right");
      pl->sublist("Periodic BCs").setParameters( *pbc );
    }
    
    mesh_factory->setParameterList(pl);
    Teuchos::RCP<panzer_stk::STK_Interface> mesh = mesh_factory->buildUncommitedMesh(*(Comm->getRawMpiComm()));
    
    mesh_factory->completeMeshConstruction(*mesh,*(Comm->getRawMpiComm()));
    
    if (Comm->getRank() == 0) {
      mesh->printMetaData(std::cout);
    }
    vector<string> blocknames;
    mesh->getElementBlockNames(blocknames);
    int spaceDim = mesh->getDimension();
    
    // ==========================================================
    // Define the connectivity manager
    // ==========================================================
    
    Teuchos::RCP<panzer::ConnManager> conn = Teuchos::rcp(new panzer_stk::STKConnManager(mesh));
    
    // ==========================================================
    // Test out a few DOF managers
    // ==========================================================
    
    {
      if (Comm->getRank() == 0) {
        std::cout << "================================================" << std::endl;
        std::cout << " *** Testing one HGRAD variable *** " << std::endl;
      }
      Teuchos::RCP<panzer::DOFManager> DOF = Teuchos::rcp(new panzer::DOFManager());
      DOF->setConnManager(conn,*(Comm->getRawMpiComm()));
      DOF->setOrientationsRequired(true);
      
      for (size_t b=0; b<blocknames.size(); b++) {
        topo_RCP cellTopo = mesh->getCellTopology(blocknames[b]);
        basis_RCP basis = Teuchos::rcp(new Intrepid2::Basis_HGRAD_HEX_C1_FEM<PHX::Device::execution_space,double,double>() );
        Teuchos::RCP<const panzer::Intrepid2FieldPattern> Pattern = Teuchos::rcp(new panzer::Intrepid2FieldPattern(basis));
        DOF->addField(blocknames[b], "T", Pattern, panzer::FieldType::CG);
      }
      
      DOF->buildGlobalUnknowns();
      if (Comm->getRank() == 0) {
        DOF->printFieldInformation(std::cout);
        std::cout << "================================================" << std::endl << std::endl;
      }
    }
    
    {
      if (Comm->getRank() == 0) {
        std::cout << "================================================" << std::endl;
        std::cout << " *** Testing one HCURL variable *** " << std::endl;
      }
      Teuchos::RCP<panzer::DOFManager> DOF = Teuchos::rcp(new panzer::DOFManager());
      DOF->setConnManager(conn,*(Comm->getRawMpiComm()));
      DOF->setOrientationsRequired(true);
      
      for (size_t b=0; b<blocknames.size(); b++) {
        topo_RCP cellTopo = mesh->getCellTopology(blocknames[b]);
        basis_RCP basis = Teuchos::rcp(new Intrepid2::Basis_HCURL_HEX_I1_FEM<PHX::Device::execution_space,double,double>() );
        Teuchos::RCP<const panzer::Intrepid2FieldPattern> Pattern = Teuchos::rcp(new panzer::Intrepid2FieldPattern(basis));
        DOF->addField(blocknames[b], "E", Pattern, panzer::FieldType::CG);
      }
      
      DOF->buildGlobalUnknowns();
      if (Comm->getRank() == 0) {
        DOF->printFieldInformation(std::cout);
        std::cout << "================================================" << std::endl << std::endl;
      }
    }
    
    {
      if (Comm->getRank() == 0) {
        std::cout << "================================================" << std::endl;
        std::cout << " *** Testing one HDIV variable *** " << std::endl;
      }
      Teuchos::RCP<panzer::DOFManager> DOF = Teuchos::rcp(new panzer::DOFManager());
      DOF->setConnManager(conn,*(Comm->getRawMpiComm()));
      DOF->setOrientationsRequired(true);
      
      for (size_t b=0; b<blocknames.size(); b++) {
        topo_RCP cellTopo = mesh->getCellTopology(blocknames[b]);
        basis_RCP basis = Teuchos::rcp(new Intrepid2::Basis_HDIV_HEX_I1_FEM<PHX::Device::execution_space,double,double>() );
        Teuchos::RCP<const panzer::Intrepid2FieldPattern> Pattern = Teuchos::rcp(new panzer::Intrepid2FieldPattern(basis));
        DOF->addField(blocknames[b], "B", Pattern, panzer::FieldType::CG);
      }
      
      DOF->buildGlobalUnknowns();
      if (Comm->getRank() == 0) {
        DOF->printFieldInformation(std::cout);
        std::cout << "================================================" << std::endl << std::endl;
      }
    }
    
    {
      if (Comm->getRank() == 0) {
        std::cout << "================================================" << std::endl;
        std::cout << " *** Testing one HVOL variable *** " << std::endl;
      }
      Teuchos::RCP<panzer::DOFManager> DOF = Teuchos::rcp(new panzer::DOFManager());
      DOF->setConnManager(conn,*(Comm->getRawMpiComm()));
      DOF->setOrientationsRequired(true);
      
      for (size_t b=0; b<blocknames.size(); b++) {
        topo_RCP cellTopo = mesh->getCellTopology(blocknames[b]);
        basis_RCP basis = Teuchos::rcp(new Intrepid2::Basis_HVOL_C0_FEM<PHX::Device::execution_space,double,double>(*cellTopo) );
        Teuchos::RCP<const panzer::Intrepid2FieldPattern> Pattern = Teuchos::rcp(new panzer::Intrepid2FieldPattern(basis));
        DOF->addField(blocknames[b], "p", Pattern, panzer::FieldType::CG);
      }
      
      DOF->buildGlobalUnknowns();
      if (Comm->getRank() == 0) {
        DOF->printFieldInformation(std::cout);
        std::cout << "================================================" << std::endl << std::endl;
      }
    }
  }
  
  Kokkos::finalize();
  
  
  int val = 0;
  return val;
}
```

Expected output on the mesh with large indices:

```
================================================

 *** Testing one HGRAD variable *** 

DOFManager Field Information: 

   Element Block = air

      "T" is field ID 0

================================================


================================================

 *** Testing one HCURL variable *** 

DOFManager Field Information: 

   Element Block = air

      "E" is field ID 0

================================================


================================================

 *** Testing one HDIV variable *** 

terminate called after throwing an instance of 'std::invalid_argument'

  what():  ../packages/tpetra/core/src/Tpetra_Map_def.hpp:801:


Throw number = 1


Throw test that evaluated to true: minAllGID_ < indexBase_


Tpetra::Map constructor (noncontiguous): Minimum global ID = -3630 over all process(es) is less than the given indexBase = 0.

```

**Comment by dridzal** [comment]
Adding @TimWildey , who provided the reproducer.

**Comment by eric-c-cyr** [comment]
Do you have a time frame?

**Comment by dridzal** [comment]
@eric-c-cyr : @rppawlo contacted us via email with a fix.  He'll push it after creating tests.

**Comment by rppawlo** [comment]
@eric-c-cyr - I sent them a patch yesterday that fixed this. I have not pushed the fix yet since I need to write a test to cover it. Will have the fix in by early next week.

**Comment by rppawlo** [comment]
This was fixed with PR #10041
